// z-builtins/lib/methods.js
// Generated by CoffeeScript 1.7.1
(function() {
  exports.plain = ["charAt", "charCodeAt", "concat", "indexOf", "lastIndexOf", "localeCompare", "match", "replace", "search", "slice", "split", "substr", "substring", "toLocaleLowerCase", "toLocaleUpperCase", "toLowerCase", "toUpperCase", "trim", "concat", "join", "slice", "indexOf", "lastIndexOf", "forEach", "every", "some", "filter", "map", "reduce", "reduceRight", "toLocaleString", 'apply', 'bind', 'call', 'toExponential', 'toFixed', 'toLocaleString', 'toPrecision', 'exec', 'test'];

  exports.mutators = ["reverse", "sort", "pop", "shift", "push", "splice", "unshift"];

  exports.globals = ['eval', 'isFinite', 'isNaN', 'parseFloat', 'parseInt', 'decodeURI', 'decodeURIComponent', 'encodeURI', 'encodeURIComponent', 'escape', 'unescape'];

  exports.methods = {
    isArray: Array.isArray,
    stringify: JSON.stringify,
    parseJSON: JSON.parse,
    logarithm: Math.log,
    inc: function(x) {
      return x + 1;
    },
    dec: function(x) {
      return x - 1;
    },
    neg: function(x) {
      return -x;
    },
    "typeof": function(x) {
      return typeof x;
    },
    del: function(x, name) {
      delete x[name];
      return x;
    },
    "in": function(prop, obj) {
      return prop in obj;
    },
    "instanceof": function(obj, type) {
      return obj instanceof type;
    },
    mult: function(x, y) {
      return x * y;
    },
    div: function(x, y) {
      return x / y;
    },
    mod: function(x, y) {
      return x % y;
    },
    add: function(x, y) {
      return x + y;
    },
    sub: function(x, y) {
      return x - y;
    },
    bitNot: function(x) {
      return ~x;
    },
    bitAnd: function(x, y) {
      return x & y;
    },
    bitOr: function(x, y) {
      return x | y;
    },
    bitLeft: function(x, y) {
      return x << y;
    },
    bitRight: function(x, y) {
      return x >> y;
    },
    bitRightFill: function(x, y) {
      return x >>> y;
    },
    gt: function(x, y) {
      return x > y;
    },
    gte: function(x, y) {
      return x >= y;
    },
    lt: function(x, y) {
      return x < y;
    },
    lte: function(x, y) {
      return x <= y;
    },
    and: function(x, y) {
      return x && y;
    },
    or: function(x, y) {
      return x || y;
    },
    not: function(x) {
      return !x;
    }
  };

  exports.maths = ['abs', 'acos', 'asin', 'atan', 'atan2', 'ceil', 'cos', 'exp', 'floor', 'max', 'min', 'pow', 'round', 'sin', 'sqrt', 'tan'];

}).call(this);
